python/llm/example/GPU/Speculative-Decoding/llama2/speculative.py

```diff
@@ -0,0 +1,142 @@
+For Maya, the future was full of promise and possibility, and she was determined to continue doing everything in her power to help create a brighter, more ethical world for everyone.
+As she """
+
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser(description='Predict Tokens using `generate()` API for Llama2 model')
+    parser.add_argument('--repo-id-or-model-path', type=str, default="meta-llama/Llama-2-7b-chat-hf",
+                        help='The huggingface repo id for the Llama2 (e.g. `meta-llama/Llama-2-7b-chat-hf` and `meta-llama/Llama-2-13b-chat-hf`) to be downloaded'
+                             ', or the path to the huggingface checkpoint folder')
+    parser.add_argument('--prompt', type=str, default=long_input,
+                        help='Prompt to infer')
+    parser.add_argument('--precision', type=str, default='bf16',
+                        help='Main model Precision')
+    parser.add_argument('--n-predict', type=int, default=128,
+                        help='Max tokens to predict')
+    parser.add_argument('--max-draft', type=int, default=8,
+                        help='Max draft')
+    parser.add_argument('--xpu', action='store_true')
+    parser.add_argument('--temperature', type=float, default=0.2,
+                        help='Model temperature')
+    parser.add_argument('--th_stop_draft', type=float, default=0.8,
+                        help='draft stop probility')
+
+    args = parser.parse_args()
+    device = 'xpu'
+    model_path = args.repo_id_or_model_path
+    max_step_draft = args.max_draft
+
+    draft_model = AutoModelForCausalLM.from_pretrained(model_path,
+                                                        load_in_4bit=True,
+                                                        optimize_model=True,
+                                                        trust_remote_code=True)
+    draft_model = draft_model.half().to(device)
+
+    print("Assistant model loaded!")
+
+    model = AutoModelForCausalLM.from_pretrained(model_path,
+                                                    optimize_model=True,
+                                                    torch_dtype=torch.float16,
+                                                    load_in_low_bit="fp16",
+                                                    trust_remote_code=True,
+                                                    use_cache=True)
+    model = model.to(device)
+    print("Target model loaded!")
+
+    tokenizer = LlamaTokenizer.from_pretrained(model_path)
+
+    print(f"Model is {model.dtype}")
+    print(f"Max Draft number {max_step_draft}")
+    print(f"Max token number {args.n_predict}")
+
+    with torch.inference_mode():
+        prompt = LLAMA2_PROMPT_FORMAT.format(prompt=args.prompt)
+        input_ids = tokenizer(prompt, return_tensors='pt').input_ids.to(model.device)
+
+        output = model.speculative_generate(input_ids=input_ids,
```

Comment:
maybe sth like below:
```python
output = model.generate(..., speculative=true)
```

Reply:
Have update a draft version of this API, would you wind taking a look again ?
If we want to modify generate, we need to patch `GenerationMixin`. This way maybe make our code more coupling with code from different versions of Transformers.